### PR TITLE
*: Use or migrate to tectonic-cluo-operator

### PR DESCRIPTION
**Pull request by dghubble**
Deploy Tectonic clusters with `quay.io/coreos/tectonic-cluo-operator:v0.1.2`

* `tectonic-cluo-operator` will manage the CLUO v0.3.1 update-operator (coordinator) deployment and update-agent daemonset
* `update-operator` will stop managing the update-agent daemonset
* `update-operator` will auto-label Container Linux nodes
* `update-agent` will node select for the CLUO `agent=true` label

Tested:

- [x] Tectonic cluster fresh install
- [x] Upgrade from a 1.6.7-tectonic.2 cluster

### CODE REVIEW

File: config.tf [hunk 1]
```diff
@@ -25,43 +25,43 @@ variable "tectonic_container_images" {
...
 
   default = {
-    addon_resizer                   = "gcr.io/google_containers/addon-resizer:2.0"
-    alertmanager                    = "quay.io/prometheus/alertmanager:v0.7.1"
-    awscli                          = "quay.io/coreos/awscli:025a357f05242fdad6a81e8a6b520098aa65a600"
-    bootkube                        = "quay.io/coreos/bootkube:v0.5.0"
-    calico                          = "quay.io/calico/node:v1.3.0"
-    calico_cni                      = "quay.io/calico/cni:v1.9.1-4-g23fcd5f"
-    config_reload                   = "quay.io/coreos/configmap-reload:v0.0.1"
-    container_linux_update_operator = "quay.io/coreos/container-linux-update-operator:v0.2.2"
```
Comment (dghubble) [inline config.tf]: Unfortunately, container_linux_update_operator was the longest name so terraform fmt reformats the whole section and Github shows it as a bunch of changes.

File: config.tf [hunk 2]
```diff
@@ -25,43 +25,43 @@ variable "tectonic_container_images" {
...
+    prometheus                   = "quay.io/prometheus/prometheus:v1.7.1"
+    prometheus_config_reload     = "quay.io/coreos/prometheus-config-reloader:v0.0.2"
+    prometheus_operator          = "quay.io/coreos/prometheus-operator:v0.11.0"
+    stats_emitter                = "quay.io/coreos/tectonic-stats:6e882361357fe4b773adbf279cddf48cb50164c1"
+    stats_extender               = "quay.io/coreos/tectonic-stats-extender:487b3da4e175da96dabfb44fba65cdb8b823db2e"
+    tectonic_channel_operator    = "quay.io/coreos/tectonic-channel-operator:0.4.1"
+    tectonic_etcd_operator       = "quay.io/coreos/tectonic-etcd-operator:v0.0.1"
+    tectonic_monitoring_auth     = "quay.io/coreos/tectonic-monitoring-auth:v0.0.1"
+    tectonic_prometheus_operator = "quay.io/coreos/tectonic-prometheus-operator:v1.4.1"
+    tectonic_cluo_operator       = "quay.io/coreos/tectonic-cluo-operator:v0.1.0"
```
Comment (dghubble) [inline config.tf]: added ^

File: modules/tectonic/assets.tf [hunk 3]
```diff
@@ -9,28 +9,28 @@ resource "template_dir" "tectonic" {
...
+    tectonic_monitoring_auth_image     = "${var.container_images["tectonic_monitoring_auth"]}"
+    prometheus_image                   = "${var.container_images["prometheus"]}"
+    prometheus_config_reload_image     = "${var.container_images["prometheus_config_reload"]}"
+    alertmanager_image                 = "${var.container_images["alertmanager"]}"
+    stats_emitter_image                = "${var.container_images["stats_emitter"]}"
+    stats_extender_image               = "${var.container_images["stats_extender"]}"
+    tectonic_channel_operator_image    = "${var.container_images["tectonic_channel_operator"]}"
+    tectonic_prometheus_operator_image = "${var.container_images["tectonic_prometheus_operator"]}"
+    tectonic_etcd_operator_image       = "${var.container_images["tectonic_etcd_operator"]}"
+    tectonic_cluo_operator_image       = "${var.container_images["tectonic_cluo_operator"]}"
```
Comment (dghubble) [inline modules/tectonic/assets.tf]: added ^

**Comment by abhinavdahiya** [comment]
@yifan-gu @aaronlevy AppVersion objects will be moved to CRDs. So probably wait on this until that makes into tectonic-installer?

**Comment by dghubble** [comment]
I'd like to get this initial use of tectonic-cluo-operator in soon to lay the foundation. We work to do on supporting more older versions and torcx additions that will appear in a rev'd tectonic-cluo-operator (ideally we just bump version here) that must be in the next Tectonic release. 

What's the ETA on your switch to CRDs?
cc @crawford @aaronlevy

**Comment by dghubble** [comment]
Note: tectonic-cluo-operator:v0.1.0 only contains "old" manifests for 1.7.1-tectonic.1. Manifests for additional older clusters still need to be added.

**Comment by sym3tri** [comment]
Naming nit:

if you expand the acronym of "tectonic-cluo-operator" you get:
"tectonic-container-linux-update-operator-operator"

**Comment by dghubble** [comment]
@sym3tri this hasn't escaped notice. We discussed it and its unfortunate, but renaming CLUO to get the tectonic name to be right isn't worth the complexity. To support migrating clusters, the name has to stay the same in some places (otherwise it requires pruning logic too) and the migrations are tricky enough imo.

There is a broader lesson here, captured in https://github.com/coreos/container-linux-update-operator/issues/98. A project should either be an application with app-specific logic **or** an operator which creates manifests, not both. It was a mistake for cluo to take on operational responsibilities and tectonic-cluo-operator should help address this and minimize the amount of legacy code cluo must maintain to support Tectonic indefinitely. With what we know today, CLUO should have stuck with klocksmith or been called the "container linux reboot coordinator" (use of "update" also gives users the false idea its the piece doing the update). Anyway, I'd prefer we not attempt a rename in the next release.

**Comment by dghubble** [comment]
Cutting a new tectonic-cluo-operator release with a new migration update strategy...

**Comment by dghubble** [comment]
I created a test payload and finished most of a migration from 1.6.7-tectonic.2 (local environment doesn't have enough resources to complete). Thanks @yifan-gu. This is waiting on a change to allow TCO to create the missing AppVersion and re-testing.

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by Quentin-M** [comment]
This migration must go in if we want to open the way of getting torcx profiles in the next release and get CL to move forward. This will be our first tectonic-x-operator based deployment.

@yifan-gu What is the state of the TCO change required for this?
@sym3tri @alexsomesan @s-urbaniak Could we get some review on this?
@^ Do you think we can get it?

**Comment by Quentin-M** [comment]
ok to test

**Comment by dghubble** [comment]
Yep, that's a good summary.

I've been working with @yifan-gu testing this out. Just waiting on a TCO change and rebase. Also, I think this is already getting tested without the "ok to test" since my privileges are sufficient.

**Comment by aaronlevy** [comment]
Top priority in this next release is getting all TPR->CRD migration work done for AppVersion objects. This must come secondary to that - or the release needs to be pushed. /cc @robszumski @mike-saparov

**Comment by Quentin-M** [comment]
@dghubble Just wanted to re-trigger - as the tests automatically rebase the PR on master.

**Comment by Quentin-M** [comment]
@aaronlevy Ack. Thank you.

**Comment by Quentin-M** [comment]
This is technically PROD-212 as I understand it.

**Comment by abhinavdahiya** [comment]
@dghubble @aaronlevy  
currently `AllResourcesCreated` smoke test ensure all the manifests were created. And since now tectonic-cluo-operator installs cluo, this test might not catch errors where installation by tectonic-cluo-operator failed.

Maybe add another test which checks all AppVersion resources have desiredVersion == currentVersion.

**Comment by dghubble** [comment]
@abhinavdahiya I'm aware, hence you see the comments about manually testing that it behaves correctly. Existing tests only cover existing expectations. Tectonic Installer has the same problem, checking for pods to be running, but there lacking a pod entirely isn't caught.

**Comment by dghubble** [comment]
_sigh_ rebasing again, until this is ready for approval

**Comment by dghubble** [comment]
ok to test

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by Quentin-M** [comment]
retest this please

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by Quentin-M** [comment]
Closing and re-opening to trigger full smokes tests as Jenkins won't recognize the label otherwise..

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
The payloads won't actually work until the TCO fix is updated (linked above if you have the privilege to see it)

**Comment by dghubble** [comment]
On Jenkins, Azure terraform apply seems broken. For bare-metal it failed to clone the repo.

**Comment by dghubble** [comment]
ok to test

**Comment by dghubble** [comment]
test upgrade from 1.6.7-tectonic.2 1.6.7-tectonic.2

**Comment by dghubble** [comment]
Each of these failures are due to unrelated flakes or testing environment problems.

**Comment by dghubble** [comment]
Moving efforts to #1736 branch against crd_migrations which seeks to collect all the required migrations together for submission.

**Pull request closed without merging**